### PR TITLE
Require LTS version of Node for Wasp

### LINKED ISSUE #442: Require LTS node version for Wasp
Currently Wasp is very flexible in the node version it requires; it can be anything after v12. There's no particular restriction on npm either.

This flexibility doesn't seem to be a major feature - we already recommend in the docs that people install nvm to manage their versions. This flexibility does lead to far more complex testing requirements.

According to this: https://nodejs.org/en/about/releases/

node v12 and v14 currently in maintenance mode. Node v16 is the active LTS version.

In addition, Node ships with particular versions of NPM. While you *can* upgrade your npm under an older version of Node, mixing these is problematic as people generally do not have such mixed combinations. Therefore I recommend we stick to the npm version that comes with Node 16, which is npm 8.

I already noticed some oddness as I did an `start` with Node v14 and then later again with Node v16 -- there was a lot of extra output on updating package-lock versions. (everything did work)

So I think we should switch to require Node v16 (along with npm 8). This requires updating the docs. The github action build pipeline should also be updated to use this node version.

### PR DESCRIPTION
**Pull request by sodic**
# Description

Everything to require users to run Wasp with a _Node_ version compatible with `16.0.0` and an _NPM_ version compatible with `8.0.0`.

Fixes #442

### CODE REVIEW

File: waspc/data/Generator/templates/server/package.json [hunk 1]
```diff
@@ -16,7 +16,8 @@
     "delay": "1000"
   },
   "engines": {
-    "node": ">={= nodeVersion =}"
+    "node": "^{= nodeVersion =}",
+    "npm": "^{= npmVersion =}"
```
Comment (sodic) [inline waspc/data/Generator/templates/server/package.json:20]: The caret `^` stands for "semantically compatible with". For example, `^16.0.0` allows all version that look like `16.x.x`

Check [NPM's docs](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#dependencies) for details.

File: web/docs/getting-started.md [hunk 2]
```diff
@@ -11,10 +11,14 @@ import useBaseUrl from '@docusaurus/useBaseUrl';
...
-### Node.js >= 12.18.0
+### Node.js compatible with 16.0.0
+```shell-session
+node -v  # must be something like 16.x.x
+```
 
+### NPM compatible with 8.0.0
 ```shell-session
-node -v  # >= v12.18.0
+npm -v  # must be something like 8.x.x
```
Comment (sodic) [inline web/docs/getting-started.md:21]: @faassen Is this what you had in mind? We now only allow Node versions that look like `16.x.x`, meaning that 17.x.x. and higher won't work. Same goes for NPM.

File: waspc/src/Wasp/Generator/WebAppGenerator.hs [hunk 3]
```diff
@@ -83,16 +104,16 @@ npmDepsForWasp =
           []
     }
 
-generateGitignore :: Generator FileDraft
-generateGitignore =
+genGitignore :: Generator FileDraft
+genGitignore =
```
Comment (sodic) [inline waspc/src/Wasp/Generator/WebAppGenerator.hs:108]: I noticed we were predominantly using `genFoo` instead of `generateFoo`, so I changed the ones I encountered to match this convention.

On a related note, there are a lot very similar functions between `WebAppGenerator` and `ServerGenerator`. How come they aren't extracted into common?

This is the list of shared functions. All of them differ in terms of `asWebAppFile` and `asServerFile`, some differ in other ways. However, I think they all essentially do the same thing and we could probably parameterize the differences:
```
genGitignore 
genNpmrc 
genNvmrc 
genPackageJson 
genReadme 
genSrcDir 
npmDepsForWasp 
```

File: waspc/src/Wasp/Generator/Common.hs [hunk 4]
```diff
@@ -14,12 +15,22 @@ data ProjectRootDir
...
+npmVersion = (8, 0, 0) -- Latest LTS version.
+
+npmVersionAsText :: String
+npmVersionAsText = versionAsText npmVersion
+
+prismaVersion :: (Int, Int, Int)
+prismaVersion = (3, 9, 1) -- Latest LTS version.
+
+prismaVersionAsText :: String
+prismaVersionAsText = versionAsText prismaVersion
```
Comment (sodic) [inline waspc/src/Wasp/Generator/Common.hs:33]: I wanted to make this consistent with the others:
- `fooVersion` - tuple of Ints
- `fooVersionAsString` - text

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -107,6 +107,11 @@ jobs:
           
       - name: Build dependencies 
         run: stack --install-ghc test --only-dependencies
+
+      - name: Set up Node
+      - uses: actions/setup-node@v2
+        with:
+          node-version: '16'
```
Comment (sodic) [inline .github/workflows/ci.yaml:114]: @shayneczyzewski and @Martinsos The e2e tests are currently passing, but I assume GitHub will automatically bump its default version of Node to 18 when it comes out.

That's why it's probably a good idea to specify which Node version the CI should use. The change I made is based on [this](https://github.com/actions/setup-node#usage), but GitHub seems to be ignoring the file (i.e., it doesn't run the workflow). Any idea why?

Comment (shayneczyzewski) [reply .github/workflows/ci.yaml:114]: Hey @sodic good thinking pinning it in CI! I believe in this case it is just a bad UI for GitHub actions. It actually did attempt to run CI, but is failing: https://github.com/wasp-lang/wasp/actions/runs/1968440172

Screen Shot 2022-03-11 at 7 58 19 AM

```
[Error: .github#L1]
every step must define a `uses` or `run` key
```

I can try to take a look after morning meetings if still stuck 👍🏻

Comment (sodic) [reply .github/workflows/ci.yaml:114]: I had a feeling something like this might be happening, thanks!

> I can try to take a look after morning meetings if still stuck 👍🏻

No worries, I'll take care of it. I was just missing diagnostics, and you showed me where to find them :)

Comment (shayneczyzewski) [reply .github/workflows/ci.yaml]: Awesome! 🚀

**Comment by shayneczyzewski** [comment]
Here is what our messaging looks like on `wasp compile` with Node 14 for those interested:
Screen Shot 2022-03-11 at 10 48 01 AM

**Comment by sodic** [comment]
> Here is what our messaging looks like on `wasp compile` with Node 14 for those interested: Screen Shot 2022-03-11 at 10 48 01 AM

Thanks for testing this out because it's actually not right :)

It should say "Was requires node ^16.0.0". I didn't notice we mention the `>=` symbol in the message , so I only changed it in the `package.json` template files. I'll remove the duplication (i.e., find a way to specify the full semver string in a single place).

File: .gitignore [hunk 6]
```diff
@@ -1,3 +1,6 @@
 .dir-locals.el
 # Local Netlify folder
-.netlify
\ No newline at end of file
+.netlify
+
+# editor realted
+.idea/
```
Comment (sodic) [inline .gitignore:6]: I think it's a good idea to add all editor-related files to the top-level `.gitignore`, the reason being that I often use the editors in other parts of the project (e.g., web, example apps).

File: waspc/data/Generator/templates/server/package.json [hunk 7]
```diff
@@ -16,7 +16,8 @@
     "delay": "1000"
   },
   "engines": {
-    "node": ">={= nodeVersion =}"
+    "node": "{= nodeSemverString =}",
```
Comment (sodic) [inline waspc/data/Generator/templates/server/package.json:19]: I decided this name was more appropriate because the string now contains a compatibility prefix (in this case, `^`).

File: waspc/src/Wasp/Generator/Common.hs [hunk 8]
```diff
@@ -11,15 +13,36 @@ import Text.Printf (printf)
 -- | Directory where the whole web app project (client, server, ...) is generated.
 data ProjectRootDir
 
+type SemanticVersion = (Int, Int, Int)
```
Comment (sodic) [inline waspc/src/Wasp/Generator/Common.hs:16]: Is this type an overkill?

File: waspc/src/Wasp/Generator/Job/Process.hs [hunk 9]
```diff
@@ -145,5 +145,5 @@ runNodeCommandAsJob fromDir command args jobType chan = do
         _ -> Nothing
 
     waspNodeRequirementMessage =
-      "Wasp requires node >= " ++ C.nodeVersionAsText ++ " ."
+      "Wasp requires node " ++ C.nodeSemverString ++ " ."
```
Comment (sodic) [inline waspc/src/Wasp/Generator/Job/Process.hs:148]: This fixes [the duplication issue @shayneczyzewski found](https://github.com/wasp-lang/wasp/pull/504#issuecomment-1065237994).

File: waspc/src/Wasp/Generator/Common.hs [hunk 10]
```diff
@@ -11,15 +13,36 @@ import Text.Printf (printf)
...
+npmVersion = (8, 0, 0) -- Latest LTS version.
+
+npmSemverString :: String
+npmSemverString = makeSemverString compatibleWith npmVersion
+
+prismaVersion :: SemanticVersion
+prismaVersion = (3, 9, 1)
+
+prismaSemverString :: String
+prismaSemverString = makeSemverString "" prismaVersion
```
Comment (sodic) [inline waspc/src/Wasp/Generator/Common.hs:42]: The prefix is empty because I decided to pin the _prisma_ version down to exactly `3.9.1` (without tolerance) to avoid any possible compatibility issues. We can consider updating it in a different PR (I believe version 3.10 is out).

File: waspc/src/Wasp/Generator/ServerGenerator.hs [hunk 11]
```diff
@@ -148,7 +149,7 @@ genNvmrc =
     C.mkTmplFdWithDstAndData
       (asTmplFile [relfile|nvmrc|])
       (asServerFile [relfile|.nvmrc|])
-      (Just (object ["nodeVersion" .= ('v' : nodeVersionAsText)]))
+      (Just (object ["nodeVersion" .= show (getMajor nodeVersion)]))
```
Comment (sodic) [inline waspc/src/Wasp/Generator/ServerGenerator.hs:152]: We only want to specify the major version here. If we specified the entire version string (i.e., `16.0.0`), our project would work only for that _exact_ version, which we probably don't want. Unfortunately, the format doesn't allow me specify a semver compatibility string (i.e., `^16.0.0`, the thing I put inside _package.json_) so this was the next best thing.

I removed the "v" because it was redundant.

File: waspc/src/Wasp/Generator/WebAppGenerator.hs (hunk 3 above)
Comment (sodic) [reply waspc/src/Wasp/Generator/WebAppGenerator.hs:108]: @Martinsos This is the duplication we mentioned on the weekly planning.

File: waspc/src/Wasp/Generator/Common.hs (hunk 10 above)
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/Common.hs:42]: I think that is a good idea 👍🏻

File: .github/workflows/ci.yaml [hunk 12]
```diff
@@ -107,6 +107,11 @@ jobs:
           
       - name: Build dependencies 
         run: stack --install-ghc test --only-dependencies
+
+      - name: Set up Node
```
Comment (Martinsos) [inline .github/workflows/ci.yaml:111]: Nice, can't believe we didn't have this so far!

File: waspc/src/Wasp/Generator/Common.hs (hunk 8 above)
Comment (Martinsos) [inline waspc/src/Wasp/Generator/Common.hs:16]: Ha, you went all in for SemanticVersion :D!

As @shayneczyzewski said, I think it is ok to use `type` to document stuff additionally. The main question becomes, do you want to go step further and introduce proper data type? Sometimes answer will be no, for practical reasons, because you don't want to deal with all the conversions and similar.

Here however, it seems to me you already introduced decent amount of complexity, so maybe it is a good moment to introduce a `newtype` instead of just `type`.

But some other observations first: `compatibleWith :: String` feels very out of context, looking just at its signature. And you have a lot of these functions that are all about semver, but are in Wasp.Generator.Common. So for start, I would take them out into Wasp.Semver (create it).

Then, I would consider naming. So you have Version vs SemverString. Thinking about them, what is the difference? Version is really SemverVersion. So we have SemverVersion vs SemverString. What is the difference? Is it also version but as a string? Or is it smth else?
I would suggest different terminology: Version and VersionBounds. Version is `1.2.3`, VersionBounds is `>= 1.2.3` or `== 1.2.3` or `^1.2.3`. I have to admit I am mostly getting the idea for this from cabal(PVP) dependency terminology, so maybe it is worth exploring more what is the terminology that npm uses.

You could go pretty fancy with types here: introduce `data Version`, then maybe `data VersionBounds`, and maybe even a `data VersionBoundOperator`, maybe even `data VersionBound`. Or maybe just part of this.

I think this will also become clearer once you extract it into a separate file and play with it there a bit.

File: waspc/src/Wasp/Generator/Common.hs (hunk 10 above)
Comment (Martinsos) [reply waspc/src/Wasp/Generator/Common.hs:42]: Maybe just stick with `prismaVersionAsText` -> then it is clear it is not a semverString/VersionBoundary, but instead it is specific version. And then later, when I want to use it, I know what I am using, I know that I am using a specific version, because that is a decision I am making at that call site, not here (hm is it? Ok, maybe this is a bit arguable. but still I think being more precise can't hurt).

File: waspc/src/Wasp/Generator/ServerGenerator.hs (hunk 11 above)
Comment (Martinsos) [reply waspc/src/Wasp/Generator/ServerGenerator.hs:152]: Hm interesting! This comment is pretty important, and likely something that future reader of this piece of code might benefit from, so I would suggest adding this comment into the code.

File: waspc/src/Wasp/Generator/WebAppGenerator.hs (hunk 3 above)
Comment (Martinsos) [reply waspc/src/Wasp/Generator/WebAppGenerator.hs:108]: Nice job renaming! Yes, later we started using `gen` but still had some `generate` left.

Regarding duplication -> You are right that we have some duplication, but I feel that probably it is not worth spending time on there right now, that we will overengineer. So I would leave it as it is for now, and let's see how painful it will be in the future.

File: web/docs/getting-started.md (hunk 2 above)
Comment (Martinsos) [reply web/docs/getting-started.md:21]: As we discussed, I think this is ok. We had `>= 12` and that became problematic, so let's try to use more conservative approach and see how that goes for us.

**Review by Martinsos — COMMENTED** [review]
Nice job @sodic !
Looks great and you also updated the docs, and renaming to `gen` is nice!
The main thing I commented on is taking one more step with SemanticVersion since you already started going in the direction of making it a standalone thing -> check that out.

File: .github/workflows/ci.yaml (hunk 5 above)
Comment (shayneczyzewski) [reply .github/workflows/ci.yaml]: They are running now @Martinsos it was a minor CI file syntax error

File: waspc/src/Wasp/Generator/Common.hs (hunk 8 above)
Comment (sodic) [reply waspc/src/Wasp/Generator/Common.hs:16]: > Here however, it seems to me you already introduced decent amount of complexity, so maybe it is a good moment to introduce a newtype instead of just type.
>
> But some other observations first: compatibleWith :: String feels very out of context, looking just at its signature. And you have a lot of these functions that are all about semver, but are in Wasp.Generator.Common. So for start, I would take them out into Wasp.Semver (create it).
> [...]
> I think this will also become clearer once you extract it into a separate file and play with it there a bit.

Sure, I'll clean it up a bit.

> Then, I would consider naming. So you have Version vs SemverString. Thinking about them, what is the difference? Version is really SemverVersion. So we have SemverVersion vs SemverString. What is the difference? Is it also version but as a string? Or is it smth else?

The idea was this:
- A `Version` is a version identifier consisting of a major, minor and patch value
- A `SemverString` is a string that specifies the compatibility. It can just be the version or have some additional symbols.
- 
So yes, `VersionBounds` or `VersionSpecifier` definitely make more sense, I'll change it.

File: waspc/src/Wasp/Generator/Common.hs (hunk 10 above)
Comment (sodic) [reply waspc/src/Wasp/Generator/Common.hs:42]: Hm, not sure about this one.

The name matches the other ones because this variable gets embedded into the template in a way that also supports specifying version bounds (and it is a version bound, albeit a strict one). I don't think the fact that the this particular version bound happen to cover only a single version (which semver communicates by not specifying a prefix) should affect the variable's name.

If we want the string `"3.9.1"` specifically (i.e., a version string with no extra semver stuff inside it), I think we should call `show version` or something similar. In other words, string representations of versions should not mix with bound representation strings.

**Comment by sodic** [comment]
@shayneczyzewski Seems like e2e checksums are failing in the CI but work for me locally.

I took a look at the failed job: https://github.com/wasp-lang/wasp/runs/5561699348?check_suite_focus=true.

It seems the problem comes from us comparing the two `.waspchecksums` files as strings, while the items get written into the JSON array in a somewhat non-deterministic way. In this case, the positions for `MainPage.js` and `waspLogo` appear to be swapped, causing the test to fail despite the checksums being correct. I got the current golden files by running `stack test :e2e-test` on my machine.

We could somehow sort the checksums before writing them into the file, but I don't think changing our program only to address the quirky behavior of our tests makes much sense. A better approach could be comparing the `.waspchecksums` file in a smarter way (i.e., by parsing json).

What do you think?

**Comment by shayneczyzewski** [comment]
> @shayneczyzewski Seems like e2e checksums are failing in the CI but work for me locally.
> 
> I took a look at the failed job: https://github.com/wasp-lang/wasp/runs/5561699348?check_suite_focus=true.
> 
> It seems the problem comes from us comparing the two `.waspchecksums` files as strings, while the items get written into the JSON array in a somewhat non-deterministic way. In this case, the positions for `MainPage.js` and `waspLogo` appear to be swapped, causing the test to fail despite the checksums being correct. I got the current golden files by running `stack test :e2e-test` on my machine.
> 
> We could somehow sort the checksums before writing them into the file, but I don't think changing our program only to address the quirky behavior of our tests makes much sense. A better approach could be comparing the `.waspchecksums` file in a smarter way (i.e., by parsing json).
> 
> What do you think?

Sorry @sodic this is something @Martinsos and I did not notice on our pairing branch. CI and local did not fail for us, until it was run on another branch. Martin has a fix for sorting `.waspchecksums` in his Cabal branch. We could pull out that fix and put it in `main` early to unblock others. We can chat tomorrow on this, apologies for the snag here. Good reminder for e2e stuff being deterministic.

File: waspc/src/Wasp/Generator/Common.hs (hunk 8 above)
Comment (Martinsos) [reply waspc/src/Wasp/Generator/Common.hs:16]: Maybe good naming is Version, VersionBounds, and BoundsSpecifier.
Version is a.b.c, VersionBounds is an expressions consisting of Versions and BoundSpecifiers, i.e. >= 1.2 && < 2.3, while VersionSpecifier is >= or <. But it can also be 1.* right? Ah upf ok in that case it is not so simple. Maybe best to just have Version vs VersionSpecifier / Bounds and that is it, not complicate it too much. Just thinking out loud, I am sure you will figure out something reasonable, probably best not to overengineer.

File: waspc/src/Wasp/Generator/Common.hs (hunk 10 above)
Comment (Martinsos) [reply waspc/src/Wasp/Generator/Common.hs:42]: Got it! I wanted to suggest that we have only prisma version and then use that everywhere, so it is explicit we are always using a specific version, not a version bound that by accident turns out to be very specific. Because I don't think we will ever want to relax it. And knowing that it is a very specific version gives us a bit more knowledge in the places where we use it.

But we can also go the way you describe, call this version bound and use it all around. It is not a big problem, we can always refactor if we figure out better approach.

TLDR: I argumented a bit but don't have super strong opinion, choose what you like better and I am ok with it!

File: waspc/src/Wasp/Generator/Job/Process.hs [hunk 13]
```diff
@@ -110,40 +112,51 @@ runNodeCommandAsJob fromDir command args jobType chan = do
...
 
-    waspNodeRequirementMessage =
-      "Wasp requires node " ++ C.nodeSemverString ++ " ."
-        ++ " Check Wasp docs for more details: https://wasp-lang.dev/docs#requirements ."
+waspNodeRequirementMessage :: String
+waspNodeRequirementMessage =
+  unwords
+    [ "Wasp requires node " ++ show C.nodeVersionBounds ++ ".",
+      "Check Wasp docs for more details: https://wasp-lang.dev/docs#requirements."
+    ]
```
Comment (sodic) [inline waspc/src/Wasp/Generator/Job/Process.hs:162]: @Martinsos 
I thought these functions/constants were standalone enough so I extracted them, reducing the size of the `where` clause. I can put them back in if you think that's a better option.

File: waspc/src/Wasp/Generator/Job/Process.hs [hunk 14]
```diff
@@ -88,14 +90,14 @@ runNodeCommandAsJob fromDir command args jobType chan = do
   case errorOrNodeVersion of
     Left errorMsg -> exitWithError (ExitFailure 1) (T.pack errorMsg)
     Right nodeVersion ->
-      if nodeVersion < C.nodeVersion
-        then
-          exitWithError
-            (ExitFailure 1)
-            (T.pack $ "Your node version is too low. " ++ waspNodeRequirementMessage)
-        else do
+      if isVersionInBounds nodeVersion C.nodeVersionBounds
```
Comment (sodic) [inline waspc/src/Wasp/Generator/Job/Process.hs:93]: @Martinsos 

This line is probably the main argument on why it probably makes sense to introduce both `SemanticVersion` and `VersionBounds` (perhaps I should rename it to `SemanticVersionBounds`?).

If we change our requirements for the node version, we don't need to worry about updating the condition here (e.g., `<` or `<=` or something else)

File: waspc/src/Wasp/SemanticVersion.hs [hunk 15]
```diff
@@ -0,0 +1,31 @@
+module Wasp.SemanticVersion
+  ( SemanticVersion (..),
+    VersionBounds (..),
+    isVersionInBounds,
+  )
+where
+
+import Text.Printf (printf)
+
+data SemanticVersion = SemanticVersion
```
Comment (sodic) [inline waspc/src/Wasp/SemanticVersion.hs:10]: There is a [package](https://hackage.haskell.org/package/semver) on Hackage that does some of the stuff I did here. 

However, it provides a lot of features we do not yet need, and does not provide all the features we do. Besides, I did not want to drag in a dependency for something as simple as this.

Anyway, I decided not to use it for now, we might go back on this decision it in the future.

File: waspc/src/Wasp/SemanticVersion.hs [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+    patch :: Int
+  }
+  deriving (Eq, Ord)
+
+instance Show SemanticVersion where
+  show (SemanticVersion mjr mnr ptc) = printf "%d.%d.%d" mjr mnr ptc
+
+data VersionBounds
+  = Exact SemanticVersion
+  | CompatibleWith SemanticVersion
```
Comment (sodic) [inline waspc/src/Wasp/SemanticVersion.hs:22]: I thought about adding comments here, but decided they're not needed. What do you think?

File: waspc/src/Wasp/Generator/Job/Process.hs [hunk 17]
```diff
@@ -110,40 +112,51 @@ runNodeCommandAsJob fromDir command args jobType chan = do
...
+  unwords
+    [ "Your node version does not match Wasp's requirements.",
+      "You are running node" ++ show nodeVersion ++ ".",
+      waspNodeRequirementMessage
+    ]
 
-    waspNodeRequirementMessage =
-      "Wasp requires node >= " ++ C.nodeVersionAsText ++ " ."
-        ++ " Check Wasp docs for more details: https://wasp-lang.dev/docs#requirements ."
+waspNodeRequirementMessage :: String
```
Comment (Martinsos) [inline waspc/src/Wasp/Generator/Job/Process.hs:157]: Ha you probably could have let this be part of `makeNodeVersionMismatchMessage` :D. But doesn't really make a difference, all good.

File: waspc/src/Wasp/SemanticVersion.hs (hunk 15 above)
Comment (Martinsos) [reply waspc/src/Wasp/SemanticVersion.hs:10]: Ha interesting. I think you made the right choice. Most complex part is dealing with bounds, and that library doesn't have it.

Comment (Martinsos) [inline waspc/src/Wasp/SemanticVersion.hs:10]: Naming wise -> due to how Haskell does imports/exports, I tend to go with one of two approaches:
1. Give such names, to everything in the module, that it is clear from just the names what we are dealing with. That might mean prefixing all with `SemanticVersion`. In this case, you assume this module might be imported unqualified in the rest of the code.
2. Give shorter names, so it is not completely clear just from them what it is about, the context. In this case, you assume that this module will be imported qualified in the rest of the code.

In this case, I would probably go with the option 2.

So your names would be:
```
Version
VersionBounds
Exact or ExactBounds
CompatibleWith or CompatibleWithBounds
isVersionInBounds
```

And in the rest of the code you would import it qualified as `Semantic` or `SV` or smth like that.

File: waspc/src/Wasp/SemanticVersion.hs [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+  ( SemanticVersion (..),
+    VersionBounds (..),
+    isVersionInBounds,
+  )
+where
+
+import Text.Printf (printf)
+
+data SemanticVersion = SemanticVersion
+  { major :: Int,
```
Comment (Martinsos) [inline waspc/src/Wasp/SemanticVersion.hs:11]: Here you say these are `Int`. But, we know they will be non-negative, right?
Sooo, you could use more exact type, https://hackage.haskell.org/package/base-4.10.0.0/docs/Numeric-Natural.html :D. I have to admit I never used it, but it is in base and it looks to fit the role! So maybe it is a fun opportunity to give it a try. Cool thing is you will know you don't have a negative number in here.

File: waspc/src/Wasp/SemanticVersion.hs (hunk 16 above)
Comment (Martinsos) [reply waspc/src/Wasp/SemanticVersion.hs:22]: I have to admit I don't know what `CompatibleWith` stands for :D.
Maybe it is not comments that we need, but better name?
What does it stand for? Fixed major version?

File: waspc/src/Wasp/Generator/Job/Process.hs (hunk 14 above)
Comment (Martinsos) [reply waspc/src/Wasp/Generator/Job/Process.hs:93]: Yeah, cool!
I commented on the naming below.

**Review by Martinsos — COMMENTED** [review]
Nice, I like this! SemanticVersion is now a proper thing. Nice job on renaming and ensuring we use stuff in the right places.

I think only a bit of naming is left to improve.

File: waspc/e2e-test/test-outputs/waspBuild-golden/waspBuild/.wasp/build/server/.nvmrc [hunk 19]
```diff
@@ -1 +1 @@
-v12.18.0
\ No newline at end of file
+16
```
[the remark this answers is not in the document]
Comment (sodic) [reply waspc/e2e-test/test-outputs/waspBuild-golden/waspBuild/.wasp/build/server/.nvmrc:1]: Yeah, it works this way too. I thought it was more simply.

Good thing you commented on this, it reminded me about the comment that [I promised to add in a previous thread](https://github.com/wasp-lang/wasp/pull/504#discussion_r827443458), but forgot.

File: waspc/src/Wasp/SemanticVersion.hs (hunk 16 above)
Comment (sodic) [reply waspc/src/Wasp/SemanticVersion.hs:22]: Yes, it stands for "backwards compatible with" (i.e., same major versions). The name it a bit lacking. I was basing my naming on [this](https://semver.org/). What do you think about calling it `BackwardsCompatibleWith`?

File: waspc/src/Wasp/SemanticVersion.hs (hunk 15 above)
Comment (sodic) [reply waspc/src/Wasp/SemanticVersion.hs:10]: This makes sense, I'll change it.

File: waspc/src/Wasp/SemanticVersion.hs (hunk 18 above)
Comment (sodic) [reply waspc/src/Wasp/SemanticVersion.hs:11]: And I was thinking the exact opposite (i.e., loosening the type). Version names can sometimes be strange, for example, `v1.4.0-beta2` or `v1.0.0-rc`.

I thought of making these things strings, but that natural number type does look interesting, I think we can have that for now, why not.

File: waspc/src/Wasp/SemanticVersion.hs (hunk 16 above)
Comment (sodic) [reply waspc/src/Wasp/SemanticVersion.hs]: I ended up adding comments and linking to the documentation, better safe than sorry :)

**Review by sodic — COMMENTED** [review]
@Martinsos 

I've implemented the suggested fixes, added tests, added docs for version bounds, and fixed a bug in the `isVersionInBounds` function. Most of the files are the same, you can focus on the new stuff:
- _SemanticVersionTest.hs
- The `nextBreakingChangeVersion` function in `Wasp.SemanticVersion`
- The docs and new way of determining backwards compatibility in `Wasp.SemanticVersion`

File: waspc/src/Wasp/SemanticVersion.hs (hunk 18 above)
Comment (Martinsos) [reply waspc/src/Wasp/SemanticVersion.hs]: I am ok with whatever you choose for now, I thought Natural numbers might be interesting :D, but you are right about the versions being all types of things!
Although I think stuff mostly gets added at the end with `-<some_string>`, so in theory the type could then be

```hs
  major :: Natural,
  minor :: Natural,
  patch :: Natural,
  suffix :: String
```

File: waspc/src/Wasp/Generator/ServerGenerator.hs [hunk 20]
```diff
@@ -129,7 +131,12 @@ genNvmrc =
     C.mkTmplFdWithDstAndData
       (asTmplFile [relfile|nvmrc|])
       (asServerFile [relfile|.nvmrc|])
-      (Just (object ["nodeVersion" .= ('v' : nodeVersionAsText)]))
+      -- We want to specify only the major version here. If we specified the
```
Comment (Martinsos) [inline waspc/src/Wasp/Generator/ServerGenerator.hs:134]: Great!

File: waspc/src/Wasp/SemanticVersion.hs [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+
+instance Show Version where
+  show (Version mjr mnr ptc) = printf "%d.%d.%d" mjr mnr ptc
+
+data VersionBounds
+  = -- | Allows only the version exactly equal to the one specified in the bounds
+    Exact Version
+  | -- | Allows changes that do not modify the leftmost non-zero digit in major.minor.patch, as described
+    -- in node semver docs: https://github.com/npm/node-semver#caret-ranges-123-025-004
+    BackwardsCompatibleWith Version
```
Comment (Martinsos) [inline waspc/src/Wasp/SemanticVersion.hs:26]: This is a nice name!

File: waspc/src/Wasp/SemanticVersion.hs (hunk 18 above)
Comment (sodic) [reply waspc/src/Wasp/SemanticVersion.hs]: I'd rather stay away from the suffix for now. Adding it would mean having to ensure the right compatibility rules in the `isVersionInBounds` function (and they get even more complicated with the suffix).

Is it OK to add the suffix if we ever end up needing such precision (after all, this is used only for our packages)?

File: waspc/src/Wasp/SemanticVersion.hs [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+  )
+where
+
+import Numeric.Natural
+import Text.Printf (printf)
+
+data Version = Version
+  { major :: Natural,
+    minor :: Natural,
+    patch :: Natural
```
Comment (Martinsos) [inline waspc/src/Wasp/SemanticVersion.hs:14]: Ha cool you are using Natural :D! I never used it, nice to see!

File: waspc/test/SemanticVersionTest.hs [hunk 23]
```diff
@@ -0,0 +1,39 @@
+module SemanticVersionTest where
+
+import Test.Tasty.Hspec
+import Wasp.SemanticVersion
+
+spec_isVersionInBounds :: Spec
+spec_isVersionInBounds = do
+  it "Correctly determines satisfied exact version bounds" $
```
Comment (sodic) [inline waspc/test/SemanticVersionTest.hs:8]: @Martinsos How do I decide whether to put a "describe" or not?

File: waspc/src/Wasp/SemanticVersion.hs [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+isVersionInBounds :: Version -> VersionBounds -> Bool
+isVersionInBounds version bounds = case bounds of
+  (BackwardsCompatibleWith reference) -> version >= reference && version < nextBreakingChangeVersion reference
+  (Exact reference) -> version == reference
+
+nextBreakingChangeVersion :: Version -> Version
+nextBreakingChangeVersion version = case version of
+  (Version 0 0 x) -> Version 0 0 (succ x)
+  (Version 0 x _) -> Version 0 (succ x) 0
+  (Version x _ _) -> Version (succ x) 0 0
```
Comment (Martinsos) [inline waspc/src/Wasp/SemanticVersion.hs:41]: Cool code :)!

**Review by Martinsos — APPROVED** [review]
Great, looks good to me!
Please check if there are any comments left outstanding and need tackling, but in general all looks good to me, feel free to merge once you are happy!

File: waspc/src/Wasp/SemanticVersion.hs (hunk 18 above)
Comment (Martinsos) [reply waspc/src/Wasp/SemanticVersion.hs]: Oh yes exactly, I wouldn't add it now! I just wanted to say that even if we have such strange packages, it might still have sense to have those three Natural fields, and then we can just add one String field.
So yes I would certainly not add it for now!

File: waspc/test/SemanticVersionTest.hs (hunk 23 above)
Comment (Martinsos) [reply waspc/test/SemanticVersionTest.hs:8]: Well, if you have a bunch of `it`s and you want to group them, then you use `describe`!
Imagine you see these tests failing -> if you would love to see some extra context for all of them, then you do a describe. Or if you have some code that they all need to share.
It is just a hierarchy thing -> "describe" are nodes, "it" is leaves.

File: waspc/src/Wasp/SemanticVersion.hs (hunk 24 above)
Comment (sodic) [reply waspc/src/Wasp/SemanticVersion.hs:41]: Thanks, but I can't take credit. That's all on Haskell :)
